File: bench/benches/closure.go
package main

func twice(f func(int64) int64) func(int64) int64 {
	return func(x int64) int64 { return f(f(x)) }
}
func adder(i int64) func(int64) int64 {
	return func(x int64) int64 { return x + i }
}

const closureN = 100000
const closureM = 1000000007

func main() {
	bench("closure", func() int64 {
		var acc int64
		for i := int64(0); i < closureN; i++ {
			acc = (acc*31 + twice(adder(i))(i)) % closureM
		}
		return acc
	})
}

File: bench/benches/closure.java
import java.util.function.LongUnaryOperator;
class Main {
    static final int N = 100000;
    static final long M = 1000000007L;
    static LongUnaryOperator twice(LongUnaryOperator f) {
        return x -> f.applyAsLong(f.applyAsLong(x));
    }
    static LongUnaryOperator adder(long i) {
        return x -> x + i;
    }
    static long work() {
        long acc = 0;
        for (int i = 0; i < N; i++) acc = (acc * 31 + twice(adder(i)).applyAsLong(i)) % M;
        return acc;
    }
    public static void main(String[] a) {
        Bench.bench("closure", Main::work);
    }
}
